### PR TITLE
Redefine the logger to avoid excessive output at the LOG_UI

### PR DESCRIPTION
**Pull request by PaulYuuu**
As we discussed at avocado-framework/avocado#4797 and #3169, the
current logger will display a lot of excessive output on the LOG_UI. These
outputs are the same as debug.log, they should only be in logs if we
don't use verbose mode, so this is unexpected behavior. Beraldo Leal
provides a good solution in #3169, the code works for me, so I
borrowed his solution to solve this problem accordingly.
    
Some VT modules will use "logging.warning" to print messages outside
the class/function, so when import these modules, the logging.RootLogger
will be initialized before the test execution, which will make the
RootLogger is out of control. Use the LOG_UI is a better choice.

Signed-off-by: Beraldo Leal <bleal@redhat.com>
Signed-off-by: Yihuang Yu <yihyu@redhat.com>

### CODE REVIEW

File: avocado_vt/options.py
Comment (PaulYuuu) [inline avocado_vt/options.py:37]: I have not idea how to name this logger is best, so I chose the file name first. Please let me know if have any concerns.

**Comment by PaulYuuu** [comment]
Ah, seems that this solution does not work on python3.9, but works on python3.6.

I am not sure if this is the bug of python3.9 itself, so set the status to "Draft" temporarily.

**Comment by PaulYuuu** [comment]
> Ah, seems that this solution does not work on python3.9, but works on python3.6.
> 
> I am not sure if this is the bug of python3.9 itself, so set the status to "Draft" temporarily.

It's my mistake. Test results are the same in python3.9 and python3.6, I missed installing a required package so the RootLogger has been initialized on the module import phage. For instances:
https://github.com/avocado-framework/avocado-vt/blob/d4c323aac396667cde94a3d22ba419647c8e6751/virttest/virsh.py#L70-L75
https://github.com/avocado-framework/avocado-vt/blob/d4c323aac396667cde94a3d22ba419647c8e6751/virttest/env_process.py#L58-L64

To avoid the RootLogger being initialized before test execution, Use the `avocado.output.LOG_UI`/`logging.getLogger('avocado.app')` is a good choice.

**Comment by PaulYuuu** [comment]
Hello @balamuruhans @pevogam @luckyh, as I mentioned above, this patch fixes two problems, redefine the logger in the first entry(options.py) and avoid the RootLogger being initialized before test execution.
Can you help to review it? Thanks in advance.

File: avocado_vt/options.py
Comment (pevogam) [inline avocado_vt/options.py:37]: It would be better if we use small caps here - it is both more standard and we might confuse LOG with one of the avocado outputs like LOG_UI and LOG_JOB. Unless you maybe mean "LOG_VT_OPTS" or something but a simple `log = logging.getLogger(...)` is better IMO.

File: virttest/env_process.py
Comment (pevogam) [inline virttest/env_process.py:62]: Can you explain what is the reason for choosing the UI output for this? I think it was only previously printed in the log files.

File: virttest/guest_agent.py
Comment (pevogam) [inline virttest/guest_agent.py:17]: Same here and below

**Review by pevogam — COMMENTED** [review]
I left a quick review mostly with questions and small suggestions, let me know what you think.

File: avocado_vt/options.py
Comment (PaulYuuu) [reply avocado_vt/options.py:37]: Okay, that makes sense for me, will update later.

File: virttest/env_process.py
Comment (PaulYuuu) [reply virttest/env_process.py:62]: Well, I just want to find a suitable logger to handle these messages, so I think highlighting those warnings in the UI is a notification for users. but it's okay for both LOG_UI and LOG_JOB.
Actually, as I discussed with @luckyh offline, a logger provided by VT is a good enhancement to handle some logs, avocado should improve the behavior of handling the RootLogger as well.
Anyway, they are another topic, but it is an optimization indeed that we can consider.

Comment (pevogam) [reply virttest/env_process.py]: I guess it could be useful on the UI side too as long as we make sure that these imports are always needed. There are some libvirt or qemu specific dependencies for instance that some of us don't want to install to begin with and it could be too noisy to look at a warning message like this on each run if you don't have to install the given dependency for your own runs.

Comment (PaulYuuu) [reply virttest/env_process.py]: Thank you @pevogam for your clarification. And after my investigation, these messages will not be saved to the log files if using other loggers, because avocado will only save logging messages after `run_test`, so the messages generated during the import phase are lost. I will fall back to use LOG_UI.

File: virttest/virt_admin.py
Comment (pevogam) [inline virttest/virt_admin.py:65]: @PaulYuuu So I guess these warnings are not included in the job log? Or where does the `avocado.app` logger's handlers point to?

Comment (PaulYuuu) [reply virttest/virt_admin.py:65]: `avocado.app` is same as `LOG_UI`
https://github.com/avocado-framework/avocado/blob/41b5626cf2d409110338486db2489d8d224f3bab/avocado/core/output.py#L34

You are right, they are not included in the job.log. Actually, before avocado is initialized, messages called by logging will not be saved to any log files if I understand correctly, even use `LOG_JOB`
https://github.com/avocado-framework/avocado/blob/41b5626cf2d409110338486db2489d8d224f3bab/avocado/core/output.py#L36

Comment (pevogam) [reply virttest/virt_admin.py:65]: > `avocado.app` is same as `LOG_UI`
> https://github.com/avocado-framework/avocado/blob/41b5626cf2d409110338486db2489d8d224f3bab/avocado/core/output.py#L34

Alright, then this is indeed better approach.

> You are right, they are not included in the job.log. Actually, before avocado is initialized, messages called by logging will not be saved to any log files if I understand correctly, even use `LOG_JOB`
> https://github.com/avocado-framework/avocado/blob/41b5626cf2d409110338486db2489d8d224f3bab/avocado/core/output.py#L36

My only remaining worry is if some of these modules affect people that don't use them. I can do some testing once we have more reviews, feel free to ping me then.

File: avocado_vt/options.py
Comment (PaulYuuu) [reply avocado_vt/options.py]: After considering it again, since it's a global variable, I will fall back to use big caps, but I will use another name.

File: avocado_vt/options.py
Comment (luckyh) [inline avocado_vt/options.py:37]: I think we can use the name `LOG` instead because it is the default logger for this module.

**Review by luckyh — COMMENTED** [review]
LGTM, otherwise

**Review by luckyh — APPROVED** [review]
Thanks.

**Comment by pevogam** [comment]
> Thanks.

Thanks @luckyh, I will do some testing today (as I was waiting for more reviews first) and also approve if the messages are rightly placed.

**Comment by luckyh** [comment]
> Thanks @luckyh, I will do some testing today (as I was waiting for more reviews first) and also approve if the messages are rightly placed.

@pevogam ok, thank you as well :)

**Comment by pevogam** [comment]
So I did some testing and my fears were correct. We disabled some messages on our side which were previously in the logs for Qemu users and were not supposed to be as we don't use libvirt type of vms:
```diff
diff --git a/virttest/utils_libvirtd.py b/virttest/utils_libvirtd.py
index df31d8a0..df6df664 100644
--- a/virttest/utils_libvirtd.py
+++ b/virttest/utils_libvirtd.py
@@ -51,7 +51,7 @@ class Libvirtd(object):
         self.service_list = []
 
         if LIBVIRTD is None:
-            logging.warning("Libvirtd service is not available in host, "
+            logging.debug("Libvirtd service is not available in host, "
                             "utils_libvirtd module will not function normally")
 
         self.service_name = "libvirtd" if not service_name else service_name
diff --git a/virttest/virsh.py b/virttest/virsh.py
index dcc15225..6b2d1056 100644
--- a/virttest/virsh.py
+++ b/virttest/virsh.py
@@ -66,8 +66,8 @@ VIRSH_COMMAND_GROUP_CACHE_NO_DETAIL = False
 try:
     VIRSH_EXEC = path.find_command("virsh")
 except path.CmdNotFoundError:
-    logging.warning("Virsh executable not set or found on path, "
-                    "virsh module will not function normally")
+    logging.debug("Virsh executable not set or found on path, "
+                  "virsh module will not function normally")
     VIRSH_EXEC = '/bin/true'
```
and the second one overlaps with the changes here. What this means is that these warnings will be made even worse now as they will flood the UI itself. @PaulYuuu What would be optimal is to not show libvirt related warnings if the users are only using qemu vms but if not then could these libvirt-related warnings at least be redirected towards the logs and not the UI?

**Comment by PaulYuuu** [comment]
Hello @pevogam, unfortunately, I can't find a solution to redirect those warnings to the log file with the current avocado version. They are out of any class/function, so when avocado loads these modules, these related codes will be executed before avocado-vt is initialized. avocado should handle the RootLogger as well and I think the root cause you are discussing in https://github.com/avocado-framework/avocado/issues/4758.

A possible way is to move/wrap them so that they do not become global calls, to avoid they will be executed when loading the module. Users should call them when they really need them, this requires other libvirt stakeholders to improve the logic.

Therefore, before solving the root cause, we should avoid `RootLogger` being initialized during avocado initializing, otherwise a lot of excessive output will be printed in LOG_UI.

**Comment by pevogam** [comment]
> Hello @pevogam, unfortunately, I can't find a solution to redirect those warnings to the log file with the current avocado version. They are out of any class/function, so when avocado loads these modules, these related codes will be executed before avocado-vt is initialized. avocado should handle the RootLogger as well and I think the root cause you are discussing in [avocado-framework/avocado#4758](https://github.com/avocado-framework/avocado/issues/4758).
> 
> A possible way is to move/wrap them so that they do not become global calls, to avoid they will be executed when loading the module. Users should call them when they really need them, this requires other libvirt stakeholders to improve the logic.
> 
> Therefore, before solving the root cause, we should avoid `RootLogger` being initialized during avocado initializing, otherwise a lot of excessive output will be printed in LOG_UI.

@beraldoleal If you agree with @PaulYuuu's conclusions here then I am fine with this and we can transfer this issue to avocado-framework/avocado#4758 as well.

**Review by beraldoleal — APPROVED** [review]
LGTM, thank you.

**Review by pevogam — APPROVED** [review]
Ok, LGTM too. Thanks for your patience @PaulYuuu.

**Comment by luckyh** [comment]
Thanks to all.

**Pull request merged**
